go-controller/pkg/factory/factory.go

```diff
@@ -48,15 +64,40 @@ type WatchFactory struct {
 }
 
 const (
-	resyncInterval = 12 * time.Hour
+	resyncInterval        = 12 * time.Hour
+	HANDLER_ALIVE  uint32 = 0
+	HANDLER_DEAD   uint32 = 1
 )
 
-func newInformer(oType reflect.Type, inf cache.SharedIndexInformer) *informer {
-	return &informer{
-		oType:    oType,
-		inf:      inf,
-		handlers: make(map[uint64]cache.ResourceEventHandler),
+func newInformer(oType reflect.Type, sharedInformer cache.SharedIndexInformer) *informer {
+	inf := &informer{
+		oType:      oType,
+		inf:        sharedInformer,
+		handlers:   make(map[uint64]*Handler),
+		removeChan: make(chan uint64),
 	}
+
+	go func() {
+		for {
+			idToRemove, ok := <-inf.removeChan
+			if !ok {
+				// Channel closed
+				break
+			}
+
+			inf.Lock()
```

Comment:
Can we just remove all informers that are tombstone dead in this loop after the informer lock is taken?

Because we may still not get rid off all deadlock possibilities, as we just introduced an unbuffered channel. This unbuffered channel will also block on subsequent write if a read has not happened.

I could draw a scenario here when we 'can' enter a deadlock when 3 pod deletes happen in a row.

Reply:
> Can we just remove all informers that are tombstone dead in this loop after the informer lock is taken?

@rajatchopra We still need some way to wake up the goroutine and take the lock and remove the handlers, so we'd need to make a buffered channel. But see below.

> Because we may still not get rid off all deadlock possibilities, as we just introduced an unbuffered channel. This unbuffered channel will also block on subsequent write if a read has not happened.

You are entirely correct that the unbuffered channel would have this problem. To address this, I think we can make it simpler. Go doesn't care how many goroutines we have, so what about just removing the handler from a goroutine in removeHandler():

```
func (wf *WatchFactory) removeHandler(objType reflect.Type, handler *Handler) error {
	inf, ok := wf.informers[objType]
	if !ok {
		return fmt.Errorf("tried to remove unknown object type %v event handler", objType)
	}

	if !atomic.CompareAndSwapUint32(&handler.tombstone, HANDLER_ALIVE, HANDLER_DEAD) {
		// Already removed
		return fmt.Errorf("tried to remove already removed object type %v event handler %d", objType, handler.id)
	}

	logrus.Debugf("sending %v event handler %d for removal", objType, handler.id)

	go func() {
		inf.Lock()
		defer inf.Unlock()
		if _, ok := inf.handlers[handler.id]; ok {
			// Remove the handler
			delete(inf.handlers, handler.id)
			logrus.Debugf("removed %v event handler %d", oType, handler.id)
		} else {
			logrus.Warningf("tried to remove unknown object type %v event handler %d", oType, handler.id)
		}
	}()

	return nil
}
```

Reply:
Cool. Much better idea.